torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1920,14 +1987,76 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
         if not self._suppress_guards_tls():
             stack = ''.join(traceback.format_list(traceback.extract_stack()[:-2]))
             if concrete_val is sympy.true:
-                self.guards.append(ShapeGuard(expr, stack))
+                guard = ShapeGuard(expr, stack)
             elif concrete_val is sympy.false:
-                self.guards.append(ShapeGuard(sympy.Not(expr), stack))
+                guard = ShapeGuard(sympy.Not(expr), stack)
             else:
-                self.guards.append(
-                    ShapeGuard(sympy.Eq(expr, concrete_val), stack))  # type: ignore[arg-type]
+                guard = ShapeGuard(sympy.Eq(expr, concrete_val), stack)  # type: ignore[arg-type]
+
+            self.guards.append(guard)
+            self.refine_ranges(guard)
+
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
+        def simplify(expr: sympy.Expr) -> sympy.Expr:
+            """
+            Simplification specialized for range refinement.
+            """
+            return self.maybe_simplify_floordiv(self.simplify(expr))
+
+        # First, try to simplify the guard.
+        # Since we do not guarantee that the function is monotonic,
+        # we set a maximum number of iterations.
+        max_iterations = 10
+        it = 0
+        expr = guard.expr
+
+        while True:
+            it += 1
+            previous = expr
+            expr = simplify(previous)
+            if expr == previous or it >= max_iterations:
+                break
+
+        # Filter the guards that:
+        #   1. are relational operations
+        #   2. have a symbol as the left-hand side
+        #   3. already have a range
+        if (
+            not isinstance(expr, sympy.Rel)
+            or not isinstance(expr.lhs, sympy.Symbol)
+            or expr.lhs not in self.var_to_range
+        ):
+            return
+
+
+        # Update the value range of the left-hand side, if the
+        # right-hand side provides a better range.
```

Comment:
oh looks like you're doing some sort of inner VR analysis, sounds good

Reply:
A summary of what is going on:
1. Filter expressions of form: `<symbol> <relop> <expr>`
2. Evaluate the value range of: `<expr>`
3. Update `<symbol>` value range if the one from (2) is better (lower or upper bound)
4. Clear the cache for `_maybe_evaluate_static`

> As @avikchaudhuri was pointing out in chat, we should only do value range updates for univariate formulas

Why? If there are any free variables on the RHS, `sympy_interp` would just get their value range, wouldn't it? I can't see why that would give incorrect results.

> I think this would eliminate the need for `var_to_guards` as well.

I don't think so. As far as I understand, we must issue the guards that were used for refining the value ranges, right? In that sense, a value range's lower and upper bound might be refined by different guards. Thus, we need to remember what are those guards.

Reply:
OK, so I agree that this is not obviously unsound (what I suggested was a more conservative version of this, your PR is going even further.) However, I am still struggling to find a simple explanation for the refinement procedure we're doing here. I'll give more specific comments inline.

Reply:
Note that point 2 and 3 here are just: if `x < f(t)` for t in som erange, and for those t we have that `a < f(t) < b` then since `<` is transitive` x < b` irrespective of `t`. Note that this simple argument also goes through if instead of `f(t)` we have `f(t1, t2, ..., tn)`. This may help with `where`, `mul` or `div` clauses.